Sources/BaggageLogging/BaggageContextLogging.swift

```diff
@@ -0,0 +1,125 @@
+// This source file is part of the Swift Baggage Context open source project
+//
+// Copyright (c) 2020 Moritz Lang and the Swift Baggage Context project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import Baggage
+import Logging
+
+/// A `BaggageContextLogging` purpose is to be adopted by frameworks which already provide a "FrameworkContext",
+/// and to such frameworks to pass their context as `HasBaggageContext`.
+public protocol BaggageContextLogging: BaggageContextCarrier {
+    /// The logger associated with this carrier.
+    ///
+    /// It should automatically populate the loggers metadata based on the `BaggageContext` associated with this context object.
+    ///
+    /// ### Implementation note
+    ///
+    /// Implementors, such as libraries or frameworks, which conform to `BaggageContextLogging` with their "Framework Context"
+    /// types, should implement this logger by wrapping the "raw" logger associated with this context with the `logger.with(BaggageContext:)` function,
+    /// which efficiently handles the bridging of baggage to metadata context values.
+    var logger: Logger { get set }
+}
+
+public struct DefaultContext: BaggageContextLogging {
+    public var baggage: BaggageContext
+    public var logger: Logger
+
+    public init(context baggage: BaggageContext, logger: Logger) {
+        self.baggage = baggage
+        self.logger = logger.with(context: baggage)
+    }
+}
+
+// ==== ----------------------------------------------------------------------------------------------------------------
+// MARK: BaggageContext (as additional Logger.Metadata) LogHandler
+
+/// Proxying log handler which adds `BaggageContext` as metadata when log events are to be emitted.
+public struct BaggageMetadataLogHandler: LogHandler {
+    var underlying: Logger
+    var context: BaggageContext
+
+    public var logLevel: Logger.Level {
+        get {
+            self.underlying.logLevel
+        }
+        set {
+            self.underlying.logLevel = newValue
+        }
+    }
+
+    public func log(level: Logger.Level,
+                    message: Logger.Message,
+                    metadata: Logger.Metadata?,
+                    source: String,
+                    file: String,
+                    function: String,
+                    line: UInt) {
+        guard self.underlying.logLevel <= level else { return }
+
+        var effectiveMetadata = self.baggageAsMetadata()
+        if let metadata = metadata {
+            effectiveMetadata.merge(metadata, uniquingKeysWith: { _, r in r })
+        }
+        self.underlying.log(level: level, message, metadata: effectiveMetadata, source: source, file: file, function: function, line: line)
+    }
+
+    public var metadata: Logger.Metadata {
+        get {
+            [:] // FIXME: or good actually?
```

Comment:
Would it make sense to return the `underlying` Logger's metadata or is that combined with the handlers metadata anyways?

Reply:
I'm a bit worried about copying... I'll give it a look though!

Reply:
We can't! Whomp whomp :< But actually I think that is fine, because there's no setter on logger for it anyway, so noone should call those I think hmm